text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,639 @@
+
+Manual annotation seems an undesirable amount of work, so we will have to
+make do with the reflection capabilities provided by NodeJS.
+
+> What about NodeJS' `__filename` magic variable? `__filename` only ever
+> represents the *current* file, i.e. the file that that symbol occurs in. We
+> wouldn't be able to write shared code in, say, `construct.ts` to obtain the
+> file name of the `Cluster` construct.
+
+From NodeJS reflection (see Appendix 1), we are able to obtain the following
+pieces of information given a construct instance:
+
+* The construct's **unqualified class name**
+* The **file path** of the JavaScript file that exports the construct, if available.
+  (If unavailable, we will recurse up its inheritance chain until we find a class with
+  a file path).
+  * From the file path of the JavaScript file, we will also be able to determine the
+    class' package (by searching for a containing `package.json` and reading
+    the information therein).
+
+We can now fill the construct identifier's components as follows:
+
+**CLASS NAME**: the construct's unqualified class name we obtain from reflection.
+
+**LIBRARY/VERSION**: `name` and `version` from the `package.json` we found based
+on the class' file path.
+
+**SUBMODULE**: Submodule names are a concept invented by us, and don't have a
+runtime representation in NodeJS. We will have to derive the submodule from
+the file path and a helper file we generate at build time. The next section
+goes into more detail.
+
+#### Submodules
+
+Especially in monocdk and the upcoming v2, disambiguating identical class names
+by submodule is going to be essential. The question is how we will determine
+submodule names from file paths, keeping a couple of goals in mind:
+
+* The submodule names reported here should preferably the same as jsii's
+  submodule names.
+* The mechanism should still work with libraries that don't have a submodule structure
+  at all, such as non-CDK team construct libraries vended by AWS like `@aws-solutions-constructs`.
+
+Some examples of what we want to derive:
+
+File name                                                              | Submodule
+-----------------------------------------------------------------------|----------
+.../node_modules/aws-cdk-lib/lib/aws-ecs/lib/cluster.js                | "aws_ecs"
+.../node_modules/aws-cdk-lib/lib/aws-eks/lib/cluster.js                | "aws_eks"
+.../node_modules/@aws-solutions-constructs/aws-lambda-sqs/lib/index.js | ""
+
+We could obtain submodule names by establishing conventions on monocdk's
+directory structure, using rules as "ignore `/lib/` and replace `-` with `_`
+to obtain the module name", for example. This reliance on convention seems
+brittle and inflexible. Instead, let's use a configuration file.
+
+#### jsii manifest
+
+What about the jsii manifest? It contains the right module names already, and
+it seems to have a `locationInModule` entry that seems to do what we need:
```

Comment:
This should go into a section called 'Alternatives Considered' so as to not muddy the proposed design.